File: julia/logit.jl
using DataFrames
using Resampling
using GLM

data_dir = "/vol/data/kaggle/abc/"

train = readtable(string(data_dir, "train.csv"))
test = readtable(string(data_dir, "test.csv"))
questions = readtable(string(data_dir, "questions.csv"))

outdata = []
for i =  [1:10]#[1:nrow(questions)]
    println(@sprintf("Working on question %s", i))
    this_question = questions[i,:]
    this_device = this_question[1,"QuizDevice"]
    this_sequence = this_question[1, "SequenceId"]
    
    X_device = train[train["Device"] .== this_device,:]
    X_other = train[train["Device"] .!= this_device,:]
    X_other = resample(X_other, nrow(X_device))
    train_X = rbind(X_device, X_other)
    train_X["y"] = int(train_X["Device"] .== this_device)
    
    mod = glm(:(y ~ X + Y + Z), train_X, Binomial())

    this_test = matrix(test[test["SequenceId"] .== this_sequence,["X", "Y", "Z"]])
    this_test = hcat(ones(size(this_test, 1), 1), this_test)

    z = this_test * coef(mod)
    
    outdata = [outdata, mean(1 / (1 + exp(-z)))]

end

    

File: R/logit.R
library(ffbase)
library(multicore)
ffload("../data/ff/train")
ffload("../data/ff/test")
print(load("../data/questions.RData"))

preds <- mclapply(1:nrow(questions), function(i) {

    cat("Working on question", i, "\n")
    
    this.question <- questions[i,]
    this.device <- this.question$QuizDevice
    this.sequence <- this.question$SequenceId
    
    train.device <- ffwhich(train, train$Device == this.device)
    train.not <- ffwhich(train, train$Device != this.device)
    n.train.device <- length(train.device)

    X <- as.data.frame(train[train.device, c("X", "Y", "Z")])
    rows <- sample(train.not[], nrow(X))
    X <- rbind(X, as.data.frame(train[rows, c("X", "Y", "Z")]))
    X$y <- rep(c(1, 0), c(n.train.device, n.train.device))

    mod <- glm(y ~ X + Y + Z, data = X, family = binomial(link = "logit"))

    this.test <- ffwhich(test, test$SequenceId == this.sequence)
    this.test <- as.data.frame(test[this.test,])

    mean(predict(mod, this.test, type = "response"))

})

save(preds, file = "../data/logit_preds.RData")
